docs/compatibility/9.3/remove-publisher-apis.md

```diff
@@ -0,0 +1,92 @@
+---
+title: "Breaking change - Removal of DockerComposePublisher, KubernetesPublisher, and AzurePublisher"
+description: "Learn about the breaking change in .NET Aspire 9.3 where publisher APIs were removed in favor of new resource types."
+ms.date: 5/12/2025
+ai-usage: ai-assisted
+ms.custom: https://github.com/dotnet/docs-aspire/issues/3373
+---
+
+# Removal of DockerComposePublisher, KubernetesPublisher, and AzurePublisher
+
+In .NET Aspire 9.3, the `AddDockerComposePublisher`, `AddKubernetesPublisher`, and `AddAzurePublisher` APIs were removed. These APIs are now replaced with new resource types that provide a more streamlined and flexible publishing experience.
+
+## Version introduced
+
+.NET Aspire 9.3
+
+## Previous behavior
+
+In .NET Aspire 9.2, the publisher API was introduced in preview, allowing the use of the following publishers:
+
+- DockerCompose
+- Kubernetes
+- Azure
+
+These publishers were added using the following methods:
+
+- <xref:Aspire.Hosting.DockerComposePublisherExtensions.AddDockerComposePublisher*>
+- <xref:Aspire.Hosting.KubernetesPublisherExtensions.AddKubernetesPublisher*>
+- <xref:Aspire.Hosting.AzurePublisherExtensions.AddAzurePublisher*>
+
+Multiple publishers could be added, and the `aspire publish` CLI command allowed users to select one for publishing.
+
+## New behavior
+
+In .NET Aspire 9.3, the publishers are now replaced with new resource types:
+
+<!-- TODO: Add xrefs when available. -->
+
+- `DockerEnvironmentResource`
+- `KubernetesEnvironmentResource`
+- `AzureEnvironmentResource`
+
+These resources include a `PublisherCallbackAnnotation` that defines their publishing behavior. The default publisher now automatically processes all resources with this annotation to generate assets. The `aspire publish` command no longer requires selecting a publisher; it uses the default publisher to handle all annotated resources.
+
+Example:
+
+```csharp
+builder.AddDockerComposeEnvironment(publisher =>
```

Comment:
What is this callback? These don't exist. These APIs are adding a resource (which takes name).

Reply:
See https://github.com/davidfowl/aspire-ai-chat-demo/compare/main...aspire-93#diff-4014fd4e71acb3b63bf6ad364854e492d2cac63cde9204d29c09ad3dfbcdf200

Reply:
I'll fix that...